### ISSUE TITLE
DashboardMiddleware is dependant on open silo connection unnecessarily

### ISSUE THREAD

**Issue by marcin-synak** [OrleansContrib/OrleansDashboard#323]
My silo is configured with `HostSelf = false` option
I am hosting the dashboard as a part of separate, larger ASP.NET Core application.

I've mapped the dashboard in `Startup.Configure` of my ASP.NET application like this (following the "TestHostSeparate" example:

```
app.Map("/Utilities/Orleans", m =>
{
    m.UseOrleansDashboard();
});
```

**What is the problem?:**
In such configuration, my ASP.NET application fails to start if connection to silo cannot be established.

I've isolated the cause to `DashboardMiddleware` class.
- Its constructor depends on `IGrainFactory` which is simply `IClusterClient` instance. It needs to be **already connected** to the silo.
- Constructor also creates new instance of `DashboardClient`.
- `DashboardClient` constructor immediately asks for some grains.
- If for any reason, this fails (silo offline, `IClusterClient` not connected etc), the whole application fails during startup.

**What would I expect:**
`DashboardMiddleware` should not depend on silo connection when it is **constructed**. It should only need it when it is **invoked**.
In short, I expect the dashboard to need the connection only when I navigate to `/Utilities/Orleans` in my application.

Would it be possible to remove all of these dependencies from `DashboardMiddleware` constructor and only create them the first time `Invoke` method is called?

Generally, it is said that putting things that can fail in constructors is a design flaw.

**Comment by koenbeuk** [comment]
@marcin-synak 

> Additionally, I've encountered an issue with `IClusterClient` object becoming unusable after failed connection

This would be a bug in Orleans, as the `IClusterClient` is guaranteed to eventually restore the connection. I don't think we need any type of special handling in this project. It may be a good idea to capture the `SiloUnavailableException` and report a 503 (Service Unavailable) http response from the dashboard request.

**Comment by koenbeuk** [comment]
I've implemented a fix for this issue before and am using it with success in one of my projects. I will tidy it up and issue a PR shortly

**Comment by koenbeuk** [comment]
> If it fails to connect at all, it reports an exception and that's it - the instance is garbage

I would argue that this not relevant in regards to this project. No where in the Orleans documentation does it deal with additional `IClusterClient` instances. There is always just the single instance and if you need to deal with initial connection issues then you can implement a retry filter (added in this [PR](https://github.com/dotnet/orleans/pull/4161) and briefly mentioned in the [docs](https://github.com/dotnet/orleans/pull/4161)).

**Comment by marcin-synak** [comment]
@koenbeuk 
Retry filter with indefinite retries helps, but still, I cannot call `client.Connect` in `Startup.cs`. My whole application cannot wait for the silo to start. Dashboard is just a tiny part of it.

That is the whole point of this issue - to initialize the connection when middleware is invoked, not constructed. Your fix just improves error control.

BTW, `IClusterClient` and `IGrainFactory` exposes the same instance here. You can cast between these two, it does not matter which you inject.

**Comment by SebastianStehle** [comment]
If you wanna fix this I would implement something like IGrainFactory provider to have the indirection you need.
